### DISCUSSION TITLE
Does the websocket API actually work?

### DISCUSSION THREAD

**Discussion by spencerkohan** [tokio-rs/axum#2283] [Q&A]
- [ x] I have looked for existing issues (including closed) about this

### Version

    
axum = { version = "0.6.20", features = ["ws", "headers"] }

### Platform

Darwin SK-Air-M2.fritz.box 22.2.0 Darwin Kernel Version 22.2.0: Fri Nov 11 02:06:26 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8112 arm64

### Description

I am trying to implement a websocket server based on the example in the 0.6.x branch.

I am using this code:

```
// server.rs

use std::sync::Arc;

use anyhow::Result;
use axum::http::{StatusCode, Uri};
use axum::{extract::State, http::Method, routing::get, Json, Router};
use lang_tools::source_rep::tokenizer::IdentifierIDMap;
use tokio::sync::RwLock;
use tower_http::cors::{Any, CorsLayer};


pub async fn start(project_model: Arc<RwLock<ProjectModel>>) -> Result<()> {

    let cors = CorsLayer::new()
        // allow `GET` and `POST` when accessing the resource
        .allow_methods([
            Method::GET,
            Method::POST,
            Method::PATCH,
            Method::CONNECT,
            Method::DELETE,
            Method::PUT,
            Method::OPTIONS,
            Method::TRACE,
            Method::HEAD,
        ])
        // allow requests from any origin
        .allow_origin(Any);

    let app = Router::new()
        .route("/", get(handler))
        .route("/ws", get(ws_handler))
        .with_state(project_model)
        .layer(cors);

    let server = axum::Server::bind(&"0.0.0.0:8080".parse().unwrap())
        .serve(app.into_make_service())
        .await?;

    Ok(server)
}


```

```
// websocket.rs

use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        ConnectInfo, TypedHeader,
    },
    response::IntoResponse,
};

use std::net::SocketAddr;

pub async fn ws_handler(
    ws: WebSocketUpgrade,
    user_agent: Option<TypedHeader<headers::UserAgent>>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
) -> impl IntoResponse {
    eprintln!("user_agent at {addr} connecting: {:#?}", addr);
    let user_agent = if let Some(TypedHeader(user_agent)) = user_agent {
        user_agent.to_string()
    } else {
        String::from("Unknown browser")
    };
    eprintln!("`{user_agent}` at {addr} connected.");

    ws.on_upgrade(move |socket| handle_socket(socket, addr))
}

async fn handle_socket(mut socket: WebSocket, who: SocketAddr) {
    match socket.send(Message::Text("hola".to_string())).await {
        Ok(()) => {}
        Err(err) => eprintln!("ERROR sending to websocket: {:?}", err),
    };
}

```

I try to connect using: `wscat -c ws://localhost:8080/ws`

## Expected behavior:

- the line `eprintln!("user_agent at {addr} connecting: {:#?}", addr);` should print something when the socket connection is attempted
- the websocket connects, and receives the text "hola" as a message

## Actual behavior:
- Nothing is printed from the server
- Connection fails with error: `error: Unexpected server response: 500`

Is this API actually supposed to be working, and is there anything I can do to debug the issue?

**Comment by davidpdrsn** [comment]
> Does the websocket API actually work?

Yes it does. Otherwise we would have caught it in the two+ years axum has been out 😅

As for your issue, I’m not sure. Currently traveling so can’t test your code. I recommend asking in the tokio discord which has axum channel.

**Comment by xd009642** [comment]
@spencerkohan Did you manage to solve this? A colleague has been struggling with this for a few days and I've just started digging into it to try and figure out what's wrong.

As an aside this is definitely an area where the axum trace logging could be improved. As a warp user I can see a breakdown of the routing decisions made and that has made debugging issues like this far simpler though for axum I can't tell what route is chosen or why the routes are rejected.

**Reply by xd009642** [reply]
And it's solved via some help on discord, you should use https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.into_make_service_with_connect_info if you want the connect info in the route. As for the lack of traces/logs to diagnose that I guess that's something else to tackle. @davidpdrsn would you be open to a PR or issue about improving the logs there?

**Reply by davidpdrsn** [reply]
We already have logs for built-in extractors returning a rejection https://docs.rs/axum/0.6.20/axum/extract/index.html#logging-rejections

**Reply by xd009642** [reply]
Oh I see we were missing the tracing feature... Being the only library in the tokio ecosystem where tracing isn't just enabled by default and we can't just use RUST_LOG to enable it was unexpected 😐
